File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& rain) {
        int n = rain.size();

        vector<int> pref(n);
        pref[0] = rain[0];
        vector<int> suff(n);
        suff[n-1] = rain[n-1];
        for(int i = 1 ; i < n ; i++) pref[i] = max(pref[i-1],rain[i]);
        for(int i = n-2 ; i >= 0 ; i--) suff[i] = max(suff[i+1],rain[i]);
        int total = 0;
        for(int i = 0 ; i < n ; i++){
            if(rain[i] < pref[i] && rain[i] < suff[i]){
                total += (min(pref[i],suff[i]))-rain[i];
            }
        }
        return total;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
from copy import copy
class Solution:
    def trap(self, heights: List[int]) -> int:
        ansarray = [0]*len(heights)
        rightarray = ansarray.copy()
        leftarray = ansarray.copy()

        i = len(heights)-1
        tempmax = heights[-1]
        while i >= 0:
            if tempmax < heights[i]:
                tempmax = heights[i]
            rightarray[i] = tempmax
            i -= 1
        j = 0   
        tempmax = heights[0]
        while j < len(heights):
            if heights[j] > tempmax:
                tempmax = heights[j]
            leftarray[j] = tempmax
            j+=1

        print("left :",leftarray)
        print("right: ",rightarray)
        for x in range(len(heights)):
            diff = min(leftarray[x],rightarray[x])
            curr = diff-heights[x]
            ansarray[x] = curr
        return sum(ansarray)
        


